In [ ]:
NAME = "Yinglu Deng"

---

# Lab 6: Skip Gram

**Please read the following instructions very carefully**

## Working on the assignment / FAQs
- **Always use the seed/random_state as *42* wherever applicable** (This is to ensure repeatability in answers, across students and coding environments) 
- The type of question and the points they carry are indicated in each question cell
- To avoid any ambiguity, each question also specifies what *value* must be set. Note that these are dummy values and not the answers
- If an autograded question has multiple answers (due to differences in handling NaNs, zeros etc.), all answers will be considered.
- You can delete the `raise NotImplementedError()`
- **Submitting the assignment** : Download the '.ipynb' file from Colab and upload it to bcourses. Do not delete any outputs from cells before submitting.
- That's about it. Happy coding!


Available software:
 - Python's Gensim module: https://radimrehurek.com/gensim/ (install using pip)
 - Sklearn’s  TSNE module in case you use TSNE to reduce dimension (optional)
 - Python’s Matplotlib (optional)

_Note: The most important hyper parameters of skip-gram/CBOW are vector size and windows size_


In [2]:
!pip install gensim
!wget -nc https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz 


import pandas as pd
import numpy as np 
import gensim



--2021-10-26 22:48:26--  https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.217.66.70
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.217.66.70|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1647046227 (1.5G) [application/x-gzip]
Saving to: ‘GoogleNews-vectors-negative300.bin.gz’

GoogleNews-vectors- 100%[===================>]   1.53G  90.8MB/s    in 17s     

2021-10-26 22:48:42 (95.0 MB/s) - ‘GoogleNews-vectors-negative300.bin.gz’ saved [1647046227/1647046227]



In [3]:
from gensim.models import KeyedVectors

model = KeyedVectors.load_word2vec_format('GoogleNews-vectors-negative300.bin.gz', binary=True)

In [4]:
word_vectors = model.wv
word_vectors.vocab

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: Call to deprecated `wv` (Attribute will be removed in 4.0.0, use self instead).
  """Entry point for launching an IPython kernel.


{'</s>': <gensim.models.keyedvectors.Vocab at 0x7f228c056cd0>,
 'in': <gensim.models.keyedvectors.Vocab at 0x7f228afba050>,
 'for': <gensim.models.keyedvectors.Vocab at 0x7f228c026850>,
 'that': <gensim.models.keyedvectors.Vocab at 0x7f228c026750>,
 'is': <gensim.models.keyedvectors.Vocab at 0x7f22895f3110>,
 'on': <gensim.models.keyedvectors.Vocab at 0x7f22890ea5d0>,
 '##': <gensim.models.keyedvectors.Vocab at 0x7f22890ea590>,
 'The': <gensim.models.keyedvectors.Vocab at 0x7f22890ea650>,
 'with': <gensim.models.keyedvectors.Vocab at 0x7f22890ea790>,
 'said': <gensim.models.keyedvectors.Vocab at 0x7f22890ea810>,
 'was': <gensim.models.keyedvectors.Vocab at 0x7f22890ea7d0>,
 'the': <gensim.models.keyedvectors.Vocab at 0x7f22890ea850>,
 'at': <gensim.models.keyedvectors.Vocab at 0x7f22890ea8d0>,
 'not': <gensim.models.keyedvectors.Vocab at 0x7f22890ea990>,
 'as': <gensim.models.keyedvectors.Vocab at 0x7f22890ea9d0>,
 'it': <gensim.models.keyedvectors.Vocab at 0x7f22890ea3d0>,
 'be': <gen

### **Q1 (1 point)** 
Find the cosine similarity between the following word pairs

- (France, England)
- (smaller, bigger)
- (England, London)
- (France, Rocket)
- (big, bigger)

In [5]:
model.most_similar(positive=['woman', 'king'], negative=['man'])

[('queen', 0.7118192911148071),
 ('monarch', 0.6189674139022827),
 ('princess', 0.5902431011199951),
 ('crown_prince', 0.5499460697174072),
 ('prince', 0.5377321243286133),
 ('kings', 0.5236844420433044),
 ('Queen_Consort', 0.5235945582389832),
 ('queens', 0.518113374710083),
 ('sultan', 0.5098593235015869),
 ('monarchy', 0.5087411999702454)]

In [6]:
word_pairs = [
              ('France', 'England') ,
              ('smaller', 'bigger'), 
              ('England', 'London'), 
              ('France', 'Rocket'),
              ('big', 'bigger')
            ]
for pair in word_pairs:
    print(model.similarity(pair[0], pair[1]))

0.39804944
0.7302272
0.43992856
0.07114174
0.68423855


In [7]:
#Replace 0 with the code / value; Do not delete this cell
similarity_pair1 = 0.39804944
similarity_pair2 = 0.7302272
similarity_pair3 = 0.43992856
similarity_pair4 = 0.07114174
similarity_pair5 = 0.68423855



In [8]:
#This is an autograded cell, do not edit/delete
print(similarity_pair1, similarity_pair2, similarity_pair3, similarity_pair4, similarity_pair5)

0.39804944 0.7302272 0.43992856 0.07114174 0.68423855


### **Q2 (1 point)** 
Write an expression to extract the vector representations of the words: 

- France
- England
- smaller
- bigger
- rocket
- big

Get only the first 5 elements for each vector representation.

In [7]:
words = ['France', 'England', 'smaller', 'bigger', 'rocket', 'big']
for word in words:
  print(len(model[word]), model[word])

300 [ 4.85839844e-02  7.86132812e-02  3.24218750e-01  3.49121094e-02
  7.71484375e-02  3.54003906e-02 -1.25976562e-01 -3.86718750e-01
 -1.31835938e-01  2.91748047e-02 -1.44531250e-01 -1.42578125e-01
  1.79687500e-01 -2.75390625e-01 -1.65039062e-01  9.32617188e-02
  1.17187500e-01  1.82617188e-01  6.10351562e-02  1.14257812e-01
  1.82617188e-01 -1.16699219e-01 -3.24707031e-02 -7.56835938e-02
  9.64355469e-03  8.59375000e-02 -2.85156250e-01 -2.55859375e-01
  3.01513672e-02  2.16796875e-01 -1.00097656e-01  2.85644531e-02
 -2.81250000e-01 -8.39843750e-02 -2.02636719e-02 -1.96289062e-01
 -4.78515625e-02  7.12890625e-02 -1.42578125e-01 -1.13525391e-02
  1.16210938e-01  7.22656250e-02  1.47460938e-01  1.50390625e-01
  1.40625000e-01  2.47070312e-01 -1.69921875e-01  7.76367188e-02
 -5.44433594e-02  1.66992188e-01 -1.45507812e-01  2.12402344e-02
 -7.51953125e-02  4.58984375e-02 -2.55859375e-01  1.49414062e-01
 -5.62500000e-01 -1.34765625e-01 -1.87500000e-01 -3.26538086e-03
  8.44726562e-02  1.3

In [8]:
#Replace 0 with the code / value to get the first 5 elements of each vector; Do not delete this cell
vector_1 = [4.85839844e-02, 7.86132812e-02, 3.24218750e-01, 3.49121094e-02, 7.71484375e-02]
vector_2 = [-1.98242188e-01, 1.15234375e-01, 6.25000000e-02, -5.83496094e-02, 2.26562500e-01]
vector_3 = [-0.05004883, 0.03417969, -0.0703125, 0.17578125, 0.00689697]
vector_4 = [-6.54296875e-02, -9.52148438e-02, -6.22558594e-02, 1.62109375e-01, 1.98974609e-02]
vector_5 = [-3.19824219e-02, 2.71484375e-01, -2.89062500e-01, -1.54296875e-01, 1.68945312e-01]
vector_6 = [0.11132812, 0.10595703, -0.07373047, 0.18847656, 0.07666016]

In [9]:
#This is an autograded cell, do not edit/delete
print(vector_1)
print(vector_2)
print(vector_3)
print(vector_4)
print(vector_5)
print(vector_6)


[0.0485839844, 0.0786132812, 0.32421875, 0.0349121094, 0.0771484375]
[-0.198242188, 0.115234375, 0.0625, -0.0583496094, 0.2265625]
[-0.05004883, 0.03417969, -0.0703125, 0.17578125, 0.00689697]
[-0.0654296875, -0.0952148438, -0.0622558594, 0.162109375, 0.0198974609]
[-0.0319824219, 0.271484375, -0.2890625, -0.154296875, 0.168945312]
[0.11132812, 0.10595703, -0.07373047, 0.18847656, 0.07666016]


### **Q3 (1 point)** 
Find the euclidean distances between the word pairs : 

- (France, England)
- (smaller, bigger)
- (England, London)
- (France, Rocket)
- (big, bigger)


In [10]:
for pair in word_pairs:
    print(np.linalg.norm(model[pair[0]] - model[pair[1]]))

3.0151067
1.8618743
2.8752837
3.892071
1.9586496


In [11]:
#Replace 0 with the code / value; Do not delete this cell
eu_dist1 = 3.0151067
eu_dist2 = 1.8618743
eu_dist3 = 2.8752837
eu_dist4 = 3.892071
eu_dist5 = 1.9586496



In [12]:
#This is an autograded cell, do not edit / delete
print(eu_dist1)
print(eu_dist2)
print(eu_dist3)
print(eu_dist4)
print(eu_dist5)


3.0151067
1.8618743
2.8752837
3.892071
1.9586496


### **Q4 (1 point)**
Time to dabble with the power of Word2Vec. Find the 2 closest words  for the following conditions:  
- (King - Man + Queen)
- (bigger - big + small)
- (waiting - wait + run)
- (Texas + Milwaukee – Wisconsin)

In [13]:
#Replace 0 with the code / value; Do not delete this cell
closest1 = model.most_similar(positive=['King', 'Queen'], negative=['Man'], topn=2)
closest2 = model.most_similar(positive=['bigger', 'small'], negative=['big'], topn=2)
closest3 = model.most_similar(positive=['waiting', 'run'], negative=['wait'], topn=2)
closest4 = model.most_similar(positive=['Texas', 'Milwaukee'], negative=['Wisconsin'], topn=2)
closest5 = 0


In [14]:
#This is an autograded cell, do not edit/delete
print(closest1)
print(closest2)
print(closest3)
print(closest4)
print(closest5)


[('Queen_Elizabeth', 0.5257916450500488), ('monarch', 0.5004087090492249)]
[('larger', 0.7402471899986267), ('smaller', 0.732999324798584)]
[('running', 0.5654535889625549), ('runs', 0.49640005826950073)]
[('Houston', 0.7767744064331055), ('Fort_Worth', 0.7270511388778687)]
0


### **Q5 (3 points)**
Using the vectors for the words in the Google News dataset, explore the semantic representation of these words through K-means clustering and explain your findings.

*Note : Since there are ~3Mil words in the vocabulary, you can downsample it to ~20-30k randomly selected words*

**Do not delete the below cell**

In [15]:
import random
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from sklearn.decomposition import PCA
word = list(model.wv.vocab) 

random.seed(42)
num = random.randint(20000, 30000)
word_selected = word[:num]  #selected 20k words
# word_selected
df = pd.DataFrame([model[w] for w in word_selected])
df['word'] = word_selected
# df.head()


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: DeprecationWarning: Call to deprecated `wv` (Attribute will be removed in 4.0.0, use self instead).
  """


In [50]:
X = df.set_index('word')
kmeans = KMeans(n_clusters = 120, random_state = 42).fit(X)
centroids, labels = kmeans.cluster_centers_, kmeans.labels_
# print(centroids)
print(labels)

[94 94 94 ... 94  4 68]


In [51]:
index = 0
index_list = []
for i in labels:
  if i == 12:
    index_list.append(index)
  index += 1

In [52]:
len(index_list)

294

In [53]:
X_copy = X.reset_index() 
for i in index_list:
  print(X_copy.word[i])

said
says
say
told
added
called
according
announced
reported
saying
believe
comment
showed
calls
heard
felt
claims
wrote
adding
appeared
noted
believes
described
reporters
concerned
believed
confirmed
claim
claimed
appears
spoke
seemed
learned
warned
stated
revealed
suggested
expressed
denied
admitted
considering
doubt
explained
aware
ruled
speaking
declared
warning
discussed
responded
telling
talked
rejected
surprised
worried
pointed
indicated
proved
quoted
argued
thinks
opposed
suggest
informed
cited
realized
acknowledged
mentioned
concluded
recommended
criticized
suggests
dismissed
predicted
recalled
questioned
argument
blamed
insisted
mention
confirm
claiming
citing
Speaking
referring
stressed
referred
testified
approached
noticed
Asked
argue
contacted
unclear
indicate
complained
explains
noting
proof
resigned
convinced
demonstrated
disclosed
optimistic
understood
repeated
commented
challenged
indication
spoken
suggesting
assured
observed
interviewed
admits
satisfied
announcing
ins

I did the k-mean clustering to the dataset and created 120 clustering. We picked cluster 12 as an example, there are 294 total words labeled 12, the cluster 12 included "said", "says", "say", "told", "added", "called", "according", "announced", "reported" and so on. They are all verbs with any tense (past tense, present tense etc.) which shows a person's actions. These words share a same or similar feature and meaning for this cluster.

### **Q6 (1 point)**
What loss function does the skipgram model use and briefly describe what this function is minimizing.

**Do not delete the below cell**

1. Cross-Entropy Loss is used in skipgram model, because we treat this as a classification problem. It is applied to compute the loss and back-propagation updates the weights (the word embeddings).

2. Processing:

  -Build the corpus vocabulary
  -Build a skip-gram [(target, context), relevancy] generator
  -Build the skip-gram model architecture
  -Train the Model
  -Get Word Embeddings



3. 𝐽(𝜃)=−(1/𝑇) * ∑𝑡=1𝑇 * ∑log𝑝(𝑤𝑡+𝑗|𝑤𝑡)

  and 𝑝(𝑤𝑡+𝑗|𝑤𝑡)=𝑒𝑥𝑝(𝑣𝑡⋅𝑢𝑇𝑡+𝑗)∑𝑣∈𝑉𝑒𝑥𝑝(𝑣𝑡𝑢𝑇𝑣) where 𝑣𝑡 represents the word embedding for middle word 𝑤𝑡 and 𝑢𝑡+𝑗 represents word embedding for context word 𝑤𝑡+𝑗.

4. We use cross entropy to minimize the 𝐽(𝜃) function.



### **Bonus Question (1 point)** 
Find at least 2 interesting word vec combinations like the ones given in Q4

**Do not delete the below cell**

In [29]:
# YOUR CODE HERE
closest_bouns = model.most_similar(positive=['Photo', 'Camera'], negative=['Swimming'], topn=2)
closest_bouns

[('photo', 0.5570032596588135), ('Photographer', 0.5424692630767822)]